src/tools/rbd_mirror/LeaderWatcher.cc

```diff
@@ -50,19 +56,329 @@ int LeaderWatcher::init() {
+    derr << "error unregistering leader watcher for " << m_oid << " object: "
+	 << cpp_strerror(r) << dendl;
+  }
+
+  Mutex::Locker locker(m_lock);
+  assert(m_stopping);
+  assert(m_leader_lock);
+
+  m_stopping = false;
+  m_leader_lock.reset();
+}
+
+void LeaderWatcher::check_leader_alive(utime_t &now, int heartbeat_interval) {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+
+  if (m_stopping || m_releasing || m_acquiring ||
+      m_leader_lock->is_lock_owner() ||
+      now < m_leader_last_heartbeat + 2 * heartbeat_interval) {
+    return;
+  }
+
+  dout(0) << "no hearbeat from the leader for more than "
+          << 2 * heartbeat_interval << " sec -- reacquiring the leader lock"
+          << dendl;
+
+  break_leader_lock();
+  m_leader_last_heartbeat = now;
+  m_leader_locker = {};
+}
+
+void LeaderWatcher::wait_transition_complete() {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+
+  while (m_acquiring || m_releasing) {
+    m_cond.Wait(m_lock);
   }
 }
 
+void LeaderWatcher::break_leader_lock() {
+  assert(m_lock.is_locked());
+  assert(!m_stopping);
+
+  if (m_leader_locker.cookie.empty()) {
+    acquire_leader_lock();
+    return;
+  }
+
+  dout(20) << dendl;
+
+  Context *ctx = create_context_callback<
+    LeaderWatcher, &LeaderWatcher::handle_break_leader_lock>(this);
+
+  m_leader_lock->break_lock(m_leader_locker, true, ctx);
+}
+
+void LeaderWatcher::handle_break_leader_lock(int r) {
+  dout(20) << dendl;
+
+  if (r < 0 && r != -ENOENT) {
+    derr << "error beaking leader lock: " << cpp_strerror(r)  << dendl;
+    return;
+  }
+
+  Mutex::Locker locker(m_lock);
+  if (m_stopping) {
+    return;
+  }
+
+  acquire_leader_lock();
+}
+
+void LeaderWatcher::acquire_leader_lock() {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+  assert(!m_stopping);
+  m_acquiring = true;
+
+  Context *ctx = create_context_callback<
+    LeaderWatcher, &LeaderWatcher::handle_acquire_leader_lock>(this);
+
+  m_leader_lock->try_acquire_lock(ctx);
+}
+
+void LeaderWatcher::handle_acquire_leader_lock(int r) {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  if (r < 0) {
+    derr << "error acquiring leader lock: " << cpp_strerror(r)
+         << dendl;
+    m_acquiring = false;
+    m_cond.Signal();
+    if (m_stopping) {
+      return;
+    }
+    get_locker();
+    return;
+  }
+
+  if (m_stopping) {
+    m_acquiring = false;
+    m_releasing = true;
+    release_leader_lock();
+    return;
+  }
+
+  init_status_watcher();
+}
+
+void LeaderWatcher::get_locker() {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+  assert(!m_stopping);
+
+  Context *ctx = create_context_callback<
+    LeaderWatcher, &LeaderWatcher::handle_get_locker>(this);
+
+  m_leader_lock->get_locker(&m_locker, ctx);
+}
+
+void LeaderWatcher::handle_get_locker(int r) {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  if (r < 0) {
+    derr << "error retrieving leader lock owner: " << cpp_strerror(r) << dendl;
```

Comment:
Nit: if -ENOENT, no need to output error message. Should you immediately retry to acquire the lock in that case?